sound/soc/sof/ipc3-control.c

```diff
@@ -381,31 +383,50 @@ static int sof_ipc3_bytes_ext_put(struct snd_sof_control *scontrol,
 		return -EINVAL;
 	}
 
-	if (copy_from_user(cdata->data, tlvd->tlv, header.length))
-		return -EFAULT;
+	/* We need to restore cdata->data in any situation where the byte control update fails.
+	 * Save it here.
+	 */
+	old_cdata = kmemdup(cdata->data, scontrol->max_size, GFP_KERNEL);
+	if (!old_cdata)
+		return -ENOMEM;
+
+	if (copy_from_user(cdata->data, tlvd->tlv, header.length)) {
+		ret = -EFAULT;
+		goto err_free;
+	}
 
 	if (cdata->data->magic != SOF_ABI_MAGIC) {
 		dev_err_ratelimited(scomp->dev, "Wrong ABI magic 0x%08x\n", cdata->data->magic);
-		return -EINVAL;
+		goto err_restore;
 	}
 
 	if (SOF_ABI_VERSION_INCOMPATIBLE(SOF_ABI_VERSION, cdata->data->abi)) {
 		dev_err_ratelimited(scomp->dev, "Incompatible ABI version 0x%08x\n",
 				    cdata->data->abi);
-		return -EINVAL;
+		goto err_restore;
 	}
 
 	/* be->max has been verified to be >= sizeof(struct sof_abi_hdr) */
 	if (cdata->data->size > scontrol->max_size - sizeof(struct sof_abi_hdr)) {
 		dev_err_ratelimited(scomp->dev, "Mismatch in ABI data size (truncated?)\n");
-		return -EINVAL;
+		goto err_restore;
 	}
 
 	/* notify DSP of byte control updates */
-	if (pm_runtime_active(scomp->dev))
-		return sof_ipc3_set_get_kcontrol_data(scontrol, true, true);
+	if (pm_runtime_active(scomp->dev)) {
+		ret = sof_ipc3_set_get_kcontrol_data(scontrol, true, true);
+		if (ret < 0)
+			goto err_restore;
+	}
```

Comment:
I had similar patch a while ago, looks like I have dropped it.

My issue was that this will cause inconsistent behavior of the bytes control depending on if the DSP is active or not.
If the DSP is active and we update a bytes control, we have recovery, if the DSP is not active we don't.

If the user pushes the broken binary before enabling the DSP then on power on we will fail and back to the original problem.

We should have another data pointer in cdata for backup.
4. If the DSP is active, do what it is done in this path,
5. If the DSP is not active, save the backup pointer.
1. When we enter to bytes ext_put, check if we have a data backup and skip the backup (keep the one we have saved)
2. if we fail, restore the backup (the one we have saved)
3. in any case, drop the backup (if an update failed, we have restored a correct one, if the update was OK, then no need to keep the backup), so next time we will make a backup again.

Reply:
I did think of that, the reason I didn't do it like that the first time around is I simply didn't find how cdata was allocated in the first place. But yes, your suggestion does sound better than what I've implemented here.

Reply:
I have made a patch like that (did not push it just yet), the problem I'm encountering is on the firmware side now (the firmware is asked to verify the new data blob _before_ the stream is started again, which causes the verification code to wrongly accept the data blob despite it being not acceptable). So _that_ needs fixing too.

There is no good way to delay the check (and restore, to think of it) all the way to the prepare() (kernel: params?) function is it?

Reply:
This is how it has been working all the way, the blobs are sent before the stream start and they can be updated runtime.

Are you saying that we could have broken blob in topology to start with and the ext_put() would succeed but we would have a failure at a later phase?

I think the only thing we can sanely do is handle the bytes put status, anything else is out of scope of this code?

Having said that, I would expect the firmware to do some check when it receives a blob at minimum to see if it makes sense or not.

Reply:
Yeah well my inbound patches will _also_ assume the topology blob is implicitly valid and doesn't need any firmware confirmation to be trusted as valid.

One thing that slowed down the testing is... well, the actual firmware side checking itself. Right now I have a local hack to have that check actually fail and that makes it so I do have a working (but very hacky) revision. I will review it internally (in NXP) and if my coworkers deem it acceptable I'll then port it to this tree and submit a new revision.

The reason I mentioned the thing in the previous patch is that in the specific situation I'm checking the firmware validation passes when the stream is inactive and fails at params/prepare stage, which is useless to me.

With some luck I'll submit the new revision later today.